lib/nerves_hub_link/client.ex

```diff
@@ -118,139 +122,116 @@ defmodule NervesHubLink.Client do
+      def handle_fwup_message({:ok, status, message}) do
+        Logger.info("FWUP SUCCESS: #{status} #{message}")
+        reboot()
+        :ok
+      end
 
-  After a successful firmware update, NervesHubLink calls this to start the
-  reboot process. It calls `c:reboot/0` if supplied or
-  `Nerves.Runtime.reboot/0`.
-  """
-  @spec initiate_reboot() :: :ok
-  def initiate_reboot() do
-    client = mod()
+      def handle_fwup_message(fwup_message) do
+        Logger.warning("Unknown FWUP message: #{inspect(fwup_message)}")
+        :ok
+      end
 
-    {mod, fun, args} =
-      if function_exported?(client, :reboot, 0),
-        do: {client, :reboot, []},
-        else: {Nerves.Runtime, :reboot, []}
+      @impl true
+      def handle_error(error) do
+        Logger.warning("[NervesHubLink] error: #{inspect(error)}")
+      end
 
-    _ = spawn(mod, fun, args)
-    :ok
-  end
+      @doc """
+      The default implementation checks if the `:reconnect_after_msec` config has been
+      configured, and is a list of values, otherwise `NervesHubLink.Backoff.delay_list/3`
+      is used with a minimum value of 1 second, maximum value of 60 seconds, and a 50% jitter.
+      """
+      @impl true
+      def reconnect_backoff() do
+        socket_config = Application.get_env(:nerves_hub_link, :socket, [])
+
+        backoff = socket_config[:reconnect_after_msec]
+
+        if is_list(backoff) do
+          backoff
+        else
+          NervesHubLink.Backoff.delay_list(1000, 60000, 0.50)
+        end
+      end
 
-  @doc """
-  This function is called internally by NervesHubLink to notify clients of fwup errors.
-  """
-  @spec handle_error(any()) :: :ok
-  def handle_error(data) do
-    _ = apply_wrap(mod(), :handle_error, [data])
-  end
+      @doc """
+      The default implementation calls `Nerves.Runtime.reboot/0` after a successful update. This
+      is useful for testing and for doing additional work like notifying users in a UI that a reboot
+      will happen soon. It is critical that a reboot does happen.
+      """
+      @impl true
+      def reboot() do
+        _ = spawn(Nerves.Runtime, :reboot, [])
+      end
 
-  @doc """
-  This function is called internally by NervesHubLink to notify clients of disconnects.
-  """
-  @spec reconnect_backoff() :: [integer()]
-  def reconnect_backoff() do
-    backoff =
-      if function_exported?(mod(), :reconnect_backoff, 0) do
-        apply_wrap(mod(), :reconnect_backoff, [])
-      else
-        nil
+      @impl true
+      def identify() do
+        Logger.info("[NervesHubLink] identifying")
       end
 
-    if is_list(backoff) do
-      backoff
-    else
-      NervesHubLink.Backoff.delay_list(1000, 60000, 0.50)
+      defoverridable NervesHubLink.Client
     end
   end
-
-  # Catches exceptions and exits
-  defp apply_wrap(mod, function, args) do
-    apply(mod, function, args)
-  catch
-    :error, reason -> {:error, reason}
-    :exit, reason -> {:exit, reason}
-    err -> err
```

Comment:
I'd actually say this is _the_ feature of the client. Thw whole purpose here was to guard the user from doing stupid things that completely severe the connection to NervesHub and leave their device in a bricked state unable to get a remote update.

Switching to the `use` model makes sense, but I think we need to find a way to have this protection still. With the proposed changes, every direct call on the client in the socket has the potential to blow up and kill it

Reply:
nerves is smart with rollbacks to previous versions, would that not possibly help?

could we do the smart logic in the areas that the methods are called?

Reply:
Nerves can be smart with rollbacks, but it's only default in `bbb`. Most systems need to implement that logic/behavior.

And yes, totally can do the smart logic in the areas. Sorry my comment wasn't totally clear but the intention was definitely "there was a reason we did this, so lets make sure we think of that reason where things are being called"

Reply:
totally get cha

maybe everywhere one of these methods are called we wrap that call instead? we could call it `trap_errors`?